File: Solution/src/com/solution/tree/LevelOrderTraversal.java
package com.solution.tree;

import com.solution.main.BTreePrinter;
import com.solution.main.ReadInput;
import com.solution.model.Tree;

public class LevelOrderTraversal {

	public void run() {
		Tree t = ReadInput.getTree("levelOrderTraversal");

		BTreePrinter.printNode(t);
		
		doPrintLevelOrderTraversal(t);
	}

	private void doPrintLevelOrderTraversal(Tree t) {

		int height = getHeight(t);
		boolean ltr = true;
		for (int i = 1; i <= height; i++) {

			printLevel(t, i, ltr);
			ltr = !ltr;
			System.out.println();
		}

	}

	private void printLevel(Tree t, int i, boolean ltr) {

		if (i == 1) {
			System.out.print(t.getData() + " ");
		} else {
			if (ltr) {
				if (t.getLeft() != null) {
					printLevel(t.getLeft(), i - 1, ltr);

				}
				if (t.getRight() != null) {
					printLevel(t.getRight(), i - 1, ltr);

				}

			} else {
				if (t.getRight() != null) {
					printLevel(t.getRight(), i - 1, ltr);

				}
				if (t.getLeft() != null) {
					printLevel(t.getLeft(), i - 1, ltr);

				}

			}
		}

	}

	private int getHeight(Tree t) {
		if (t == null) {
			return 0;
		}
		int lheight = getHeight(t.getLeft());
		int rheight = getHeight(t.getRight());

		return Math.max(lheight, rheight) + 1;
	}

}

File: src/main/kotlin/com/cssolutions/dsalgo/tree/LevelOrderTraversal.kt
package com.cssolutions.dsalgo.tree

import com.cssolutions.dsalgo.main.BTreePrinter.printNode
import com.cssolutions.dsalgo.main.ReadInput.getTree
import com.cssolutions.dsalgo.model.Tree

class LevelOrderTraversal {
    fun run() {
        val t = getTree("levelOrderTraversal")
        printNode<Comparable<*>>(t)
        doPrintLevelOrderTraversal(t)
    }

    private fun doPrintLevelOrderTraversal(t: Tree?) {
        val height = getHeight(t)
        var ltr = true
        for (i in 1..height) {
            printLevel(t, i, ltr)
            ltr = !ltr
            println()
        }
    }

    private fun printLevel(t: Tree?, i: Int, ltr: Boolean) {
        if (i == 1) {
            print(t!!.data.toString() + " ")
        } else {
            if (ltr) {
                if (t!!.left != null) {
                    printLevel(t.left, i - 1, ltr)
                }
                if (t.right != null) {
                    printLevel(t.right, i - 1, ltr)
                }
            } else {
                if (t!!.right != null) {
                    printLevel(t.right, i - 1, ltr)
                }
                if (t.left != null) {
                    printLevel(t.left, i - 1, ltr)
                }
            }
        }
    }

    private fun getHeight(t: Tree?): Int {
        if (t == null) {
            return 0
        }
        val lheight = getHeight(t.left)
        val rheight = getHeight(t.right)
        return Math.max(lheight, rheight) + 1
    }
}